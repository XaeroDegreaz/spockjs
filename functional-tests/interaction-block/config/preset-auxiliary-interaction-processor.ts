import { InteractionProcessor } from '@spockjs/config';

export const assertionPostProcessors = [];

export const interactionProcessors: InteractionProcessor[] = [
  t => ({
    primary: false,
    declare: () => t.emptyStatement(),
    verify: () => t.emptyStatement(),
  }),
];
