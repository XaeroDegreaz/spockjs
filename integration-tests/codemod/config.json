{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert"
      }
    ]
  ]
}
