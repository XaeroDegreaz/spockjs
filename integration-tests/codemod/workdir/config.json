{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert",
        "presets": ["@spockjs/preset-sinon-mocks"]
      }
    ]
  ]
}
