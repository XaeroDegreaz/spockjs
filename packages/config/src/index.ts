import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Hooks } from './hooks';

/**
 * Config as it can be provided by the user.
 * Comments indicate transformations made to conform to InternalConfig format.
 */
export const configType = t.exact(
  t.partial({
    powerAssert: t.boolean,
    autoImport: t.union([t.boolean, t.string]),
    staticTruthCheck: t.boolean,
    assertFunctionName: t.string,
    presets: t.array(t.string),
  }),
  'Config',
);
export type Config = t.TypeOf<typeof configType>;

/**
 * Config that disables all but the core set of features.
 */
export const minimalConfig: Config = {
  powerAssert: false,
  autoImport: false,
};

/**
 * Config containing the settings used if they are missing in the user-specified config.
 */
export const defaultConfig = {
  powerAssert: true,
  autoImport: true,
  staticTruthCheck: false,
  assertFunctionName: '',
  presets: [],
};

/**
 * Config after sanitizing in extractConfigFromState.
 */
export interface InternalConfig {
  readonly powerAssert: boolean;
  readonly autoImport: string; // empty => no auto import
  readonly staticTruthCheck: boolean;
  readonly assertFunctionName: string; // empty => no custom assert function name
  readonly hooks: Hooks;
}

export const extractConfigFromState = ({
  opts,
}: {
  opts: any;
}): InternalConfig => {
  const decoded = configType.decode(opts);

  let {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    presets,
  } = decoded.getOrElseL(errors => {
    throw new Error(
      ['Invalid plugin configuration:', ...failure(errors)].join('\n'),
    );
  });

  // powerAssert
  if (powerAssert === undefined) {
    ({ powerAssert } = defaultConfig);
  }

  // autoImport
  if (autoImport === undefined) {
    ({ autoImport } = defaultConfig);
  }
  if (autoImport === false) {
    autoImport = '';
  }
  if (autoImport === true) {
    autoImport = 'power-assert';
  }

  // staticTruthCheck
  if (staticTruthCheck === undefined) {
    ({ staticTruthCheck } = defaultConfig);
  }

  // assertFunctionName
  if (assertFunctionName === undefined) {
    ({ assertFunctionName } = defaultConfig);
  }

  // presets
  if (presets === undefined) {
    ({ presets } = defaultConfig);
  }
  const hooks = presets.map(preset => require(preset) as Hooks).reduce(
    (accHooks, moreHooks) => ({
      assertionPostProcessors: [
        ...accHooks.assertionPostProcessors,
        ...moreHooks.assertionPostProcessors,
      ],
    }),
    { assertionPostProcessors: [] },
  );

  return {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    hooks,
  };
};
