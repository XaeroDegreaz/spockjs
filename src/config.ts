export interface Config {
  powerAssert: boolean;
}

const defaultConfig: Config = {
  powerAssert: true,
};

export default defaultConfig;
