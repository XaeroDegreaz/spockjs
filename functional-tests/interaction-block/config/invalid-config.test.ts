import { transform } from '@babel/core';

import plugin from '@spockjs/babel-plugin-spock';
import { Config, minimalConfig } from '@spockjs/config';

// mark implicit dependency for jest
() => require('./preset-auxiliary-interaction-processor');
const presetAuxiliaryInteractionProcessor = require.resolve(
  './preset-auxiliary-interaction-processor',
);

test('throws if the config contains no preset defining a primary interaction processor', () => {
  expect(() =>
    transform(`stub: s() >> 42;`, {
      plugins: [
        [
          plugin,
          {
            ...minimalConfig,
            presets: [presetAuxiliaryInteractionProcessor],
          } as Config,
        ],
      ],
      filename: 'test.js',
    }),
  ).toThrowErrorMatchingSnapshot();
});
