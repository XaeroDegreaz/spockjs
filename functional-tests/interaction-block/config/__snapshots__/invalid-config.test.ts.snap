// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if the config contains no preset defining a primary interaction processor 1`] = `
"Found an interaction declaration, but no preset defines a primary interactionProcessor.
You need to enable a preset for your mocking library."
`;
